btx/diagnostics/bayesian_optimization.py

```diff
@@ -0,0 +1,413 @@
+Author: Paul-Emile Giacomelli
+Date: Fall 2023
+Description: Bayesian optimization logic to be implemented in Airflow Branch Operator and Slurm task
+"""
+
+
+# Imports
+import math
+import numpy as np
+from sklearn.gaussian_process import GaussianProcessRegressor
+from sklearn.gaussian_process.kernels import RBF
+from scipy.stats import norm
+import os
+import csv
+import yaml
+import shutil
+
+
+class BayesianOptimization:
+
+    def __init__(self, criterion_name, first_loop_task, exit_loop_task, max_iterations=None):
+        if criterion_name == "max_iterations":
+                self.iteration = 0
+                self.max_iterations = max_iterations
+        else:
+            raise KeyError(f"The stop criterion name {criterion_name} does not exist.")
+        self.criterion_name = criterion_name
+        self.first_loop_task = first_loop_task
+        self.exit_loop_task = exit_loop_task
+
+    # Initial samples
+
+    def random_param_samples(self, config):
+        """
+        Function called by the "bo_init_samples_configs" task.
+
+        Creates a matrix with each row containing a set of random parameters.
+        Each parameter value falls into the range that has been defined in the config.
+
+        Parameters
+        ----------
+        config : AttrDict
+            The current config.
+        """
+        params_ranges_keys = [key for key in config if key.startswith("range_")]
+        n_params = len(params_ranges_keys)
+        n_samples = config.bayesian_opt.n_samples_init
+        n_points = config.bayesian_opt.n_points_per_param
+
+        random_param_samples = np.empty(shape=(n_samples, n_params))
+
+        for i in range(n_params):
+            # Get the parameter range
+            param_min, param_max = config["bayesian_opt"][params_ranges_keys[i]]
+
+            # Get n_samples random values of the parameter within its range 
+            param_range = np.linspace(param_min, param_max, n_points)
+            param_sample = np.random.choice(param_range, size=n_samples)
+
+            random_param_samples[:, i] = param_sample
+
+        # Return a matrix with each row containing a set of random parameters
+        return random_param_samples
+    
+        
+    # Acquisition functions
+
+    def expected_improvement(self, X, gp_model, best_y):
+        y_pred, y_std = gp_model.predict(X, return_std=True)
+        z = (y_pred - best_y) / y_std
+        ei = (y_pred - best_y) * norm.cdf(z) + y_std * norm.pdf(z)
+        return ei
+
+    def upper_confidence_bound(self, X, gp_model, beta):
+        y_pred, y_std = gp_model.predict(X, return_std=True)
+        ucb = y_pred + beta * y_std
+        return ucb
+
+    def probability_of_improvement(self, X, gp_model, best_y):
+        y_pred, y_std = gp_model.predict(X, return_std=True)
+        z = (y_pred - best_y) / y_std
+        pi = norm.cdf(z)
+        return pi
+    
+    
+    # Bayesian Optimization
+
+    @classmethod
+    def run_bayesian_opt(self, config):
```

Comment:
For future reference, class methods typically take `cls` as the first parameter instead of `self` just for clarity. This is just for legibility, the methods will work as they're written.

Reply:
I use "cls" now, thank you for the advice!